test/ExecutionFramework.cpp

```diff
@@ -50,25 +55,294 @@ string getIPCSocketPath()
+			"/usr/local/lib/libevmone.so",
+			// TODO the circleci docker image somehow only has the .a file
+			"/usr/lib/libevmone.a"
+		})
+		{
+			evmc_loader_error_code errorCode = {};
+			evmc_instance* vm = evmc_load_and_create(path, &errorCode);
+			if (!vm || errorCode != EVMC_LOADER_SUCCESS)
+				continue;
+			theVM = make_unique<evmc::vm>(vm);
+			break;
+		}
+		BOOST_REQUIRE(theVM);
+	}
+	return *theVM;
+}
+
+Address convertFromEVMC(evmc_address const& _addr)
+{
+	return Address(bytes(begin(_addr.bytes), end(_addr.bytes)));
+}
+evmc_address convertToEVMC(Address const& _addr)
+{
+	evmc_address a;
+	for (size_t i = 0; i < 20; ++i)
+		a.bytes[i] = _addr[i];
+	return a;
+}
+h256 convertFromEVMC(evmc_bytes32 const& _data)
+{
+	return h256(bytes(begin(_data.bytes), end(_data.bytes)));
+}
+evmc_bytes32 convertToEVMC(h256 const& _data)
+{
+	evmc_bytes32 d;
+	for (size_t i = 0; i < 32; ++i)
+		d.bytes[i] = _data[i];
+	return d;
+}
+
+
 }
 
+class ExecutionFramework::Host: public evmc::Host
+{
+public:
+	explicit Host(evmc::vm& _vm, langutil::EVMVersion _evmVersion):
+		m_vm(_vm)
+	{
+		if (_evmVersion == langutil::EVMVersion::homestead())
+			m_evmVersion = EVMC_HOMESTEAD;
+		else if (_evmVersion == langutil::EVMVersion::tangerineWhistle())
+			m_evmVersion = EVMC_TANGERINE_WHISTLE;
+		else if (_evmVersion == langutil::EVMVersion::spuriousDragon())
+			m_evmVersion = EVMC_SPURIOUS_DRAGON;
+		else if (_evmVersion == langutil::EVMVersion::byzantium())
+			m_evmVersion = EVMC_BYZANTIUM;
+		else if (_evmVersion == langutil::EVMVersion::constantinople())
+			m_evmVersion = EVMC_CONSTANTINOPLE;
+		else if (_evmVersion == langutil::EVMVersion::petersburg())
+			m_evmVersion = EVMC_PETERSBURG;
+		else
+			BOOST_REQUIRE(false);
+	}
+
+	struct Account
+	{
+		evmc_uint256be balance = {};
+		size_t nonce = 0;
+		bytes code;
+		evmc_bytes32 codeHash = {};
+		std::map<evmc_bytes32, evmc_bytes32> storage;
+	};
+
+	struct State
+	{
+		std::map<evmc_address, Account> accounts;
+		std::vector<LogEntry> logs;
+	};
+
+	Account* account(evmc_address const& _address)
+	{
+		auto it = m_state.accounts.find(_address);
+		return it == m_state.accounts.end() ? nullptr : &it->second;
+	}
+
+	void reset() { m_state = State{}; m_currentAddress = {}; }
+
+	bool account_exists(evmc_address const& _addr) noexcept final
+	{
+		return account(_addr) != nullptr;
+	}
+
+	evmc_bytes32 get_storage(evmc_address const& _addr, evmc_bytes32 const& _key) noexcept final
+	{
+		if (Account* acc = account(_addr))
+			return acc->storage[_key];
+		return {};
+	}
+
+	evmc_storage_status set_storage(
+		evmc_address const& _addr,
+		evmc_bytes32 const& _key,
+		evmc_bytes32 const& _value
+	) noexcept final
+	{
+		evmc_bytes32 previousValue = m_state.accounts[_addr].storage[_key];
+		m_state.accounts[_addr].storage[_key] = _value;
+
+		// TODO there are more possible values, especially "MODIFIED_AGAIN".
+		return previousValue == _value ? EVMC_STORAGE_UNCHANGED : EVMC_STORAGE_MODIFIED;
```

Comment:
This is because of the "net sstore cost". For non-Constantinople you can just report MODIFIED and ADDED. Full support for Constantinople is complicated because you have to track original storage values before transaction started.

https://github.com/ethereum/evmone/blob/master/test/utils/host_mock.hpp#L49-L68